compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java

```diff
@@ -178,6 +194,54 @@ public ComplexMatchResult multiplyAddSub(BinaryNode binary, ValueNode a, ValueNo
         return builder -> getArithmeticLIRGenerator().emitMSub(operand(a), operand(b), operand(c));
     }
 
+    /**
+     * If x < 0:
+     *
+     * -> -((-x) & 1) == 0.
+     *
+     * else:
+     *
+     * -> ((((x >> (bitCount - 1)) >>> (bitCount - 1)) + x) & (-2)) == x.
+     */
+    @MatchRule("(If (IntegerEquals (Negate=cmpLeft (And (Negate value) Constant=a)) Constant=b))")
+    @MatchRule("(If (IntegerEquals (And=cmpLeft (Add (UnsignedRightShift (RightShift value Constant=a) Constant=a) value) Constant=b) value))")
```

Comment:
I think this complex rules will hardly match any real world programs. Especially since simplifications may be done on a higher level in the compiler. I would vote for removing this special case.

Reply:
Actually, most of  the rem for power-of-2 computations will generate this pattern  `((((x >> (bitCount - 1)) >>> (bitCount - 1)) + x) & (-2)) == x` . So I think this rule is needed just for the odd/even checking.

Reply:
Indeed. The issue here is, that it relies on the canonicalization happening `SignedRemNode.canonical`. But overall, this optimization is a generic optimization we should introduce, because other platforms can also benefit from this, when the result of the modulo is compared against zero.
I am thinking of `if(x%2==0)` could be transformed to `if(x&((1<<1)-1)==0)`. I'll go and implement this kind of optimization in the HIR, so we can focus on the simpler cases in the code generation.